mod/dagmodifier.go

```diff
@@ -627,3 +647,315 @@ func (dm *DagModifier) GetCtx() context.Context {
+		return nil, err
+	}
+
+	newRoot, err := mdm.GetDb().AttachMetadataDag(dnode, fileSize, newMetaNode)
+	if err != nil {
+		return nil, err
+	}
+	mdm.GetDb().Add(newRoot)
+
+	return newRoot, nil
+}
+
+// RemoveMetadata first truncate the metadata dag from the given `root`,
+// update the metadata map with the given `metadata`,
+// then calls meta-dag builder to create a metadata subDag and
+// attach the metadata dag to a new root with the given `dataroot` being
+// a child of the new root.
+// The preconditions include that first the given `metadata` is a string with keys separated by
+// commas, second`mdm.curNode` has the root of the metadata sub-DAG.
+func (mdm *MetaDagModifier) RemoveMetadata(root ipld.Node, metakeys []byte) (ipld.Node, error) {
+	// Read the existing metadata map.
+	b, err := mdm.readMetadataBytes(root)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine the specific scenario.
+	// Scenario #1:
+	var newMetaNode ipld.Node
+	var children *ft.DagMetaNodes
+	var modificationRequired bool
+	var clear bool
+	children, err = ft.GetChildrenForDagWithMeta(mdm.ctx, root, mdm.dagserv)
+	if err != nil {
+		return nil, err
+	}
+	if children == nil {
+		return nil, errors.New("expected DAG node with metadata child node")
+	}
+
+	if b != nil {
+
+		// Create existing map and input metadata map and check.
+		m := make(map[string]interface{})
+		err = json.Unmarshal(b, &m)
+		if err != nil {
+			return nil, err
+		}
+
+		inputKeys := strings.Split(string(metakeys), ",")
+
+		exists := keyIntersects(m, inputKeys)
+
+		if exists {
+			modificationRequired = true
+			// Check if inputM and m have the same key set
+			if equalKeySets(m, inputKeys) {
+				// Scenario #1: clear metadata.
+				clear = true
+			} else {
+				// Scenario #2: delete a subset of the metadata map.
+				// iterate the inputKeys to delete each key from the existing map.
+				for _, k := range inputKeys {
+					delete(m, k)
+				}
+
+				b, err = json.Marshal(m)
+				if err != nil {
+					return nil, err
+				}
+				// Create a metadata sub-DAG
+				newMetaNode, err = mdm.buildNewMetaDataDag(b)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	if !modificationRequired {
+		return root, nil
+	}
+
+	dnode := children.DataNode
+	// if clear is true, return dnode as root.
+	if clear {
+		return dnode, nil
+	}
+
+	// Attach the modified metadata sub-DAG to a new root for the BTFS file DAG.
+	fileSize, err := FileSize(dnode)
+	if err != nil {
+		return nil, err
+	}
+
+	newRoot, err := mdm.GetDb().AttachMetadataDag(dnode, fileSize, newMetaNode)
+	if err != nil {
+		return nil, err
+	}
+	mdm.GetDb().Add(newRoot)
+
+	return newRoot, nil
+}
+
+func (mdm *MetaDagModifier) buildNewMetaDataDag(metaBytes []byte) (ipld.Node, error) {
+	var superMeta help.SuperMeta
+	err := json.Unmarshal(metaBytes, &superMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	r := bytes.NewReader(metaBytes)
+	if superMeta.ChunkSize == 0 {
+		superMeta.ChunkSize = 256 * 1024
```

Comment:
read this from a constant, I think it's in `go-btfs-chunker/splitting.go:17`

Reply:
Done.